src/scene-entry-manager.js

```diff
@@ -220,12 +230,24 @@ export default class SceneEntryManager {
     });
 
     this.scene.addEventListener("pinned", e => {
-      const el = e.detail.el;
-      const networkId = el.components.networked.data.networkId;
-      const gltfNode = pinnedEntityToGltf(el);
-      el.setAttribute("networked", { persistent: true });
-
-      this.hubChannel.pin(networkId, gltfNode);
+      if (this.authChannel.authenticated) {
+        this._pinElement(e.detail.el);
+      } else {
+        const wasInVR = this.scene.is("vr-mode");
+        if (wasInVR) this.scene.exitVR();
+        const continueTextId = wasInVR ? "entry.return-to-vr" : "dialog.close";
+
+        this.onRequestAuthentication("sign-in.pin", "sign-in.pin-complete", continueTextId, () => {
+          if (this.authChannel.authenticated) {
+            this._pinElement(e.detail.el);
+          } else {
+            // UI pins the entity optimistically, so we undo that here.
```

Comment:
it seems like this logic belongs in the pin button onclick (on grab, really) handler, and we should gate there and have no side effects if you click it and are signed in

Reply:
Yeah, I wasn't sure about all of this since all our pinning logic is event-based and basically "fire and forget". I'll try to improve it but I may have to rework a lot of the existing code.
Also not sure what the correct way is to use "manager" code like this all the way inside an a-frame component, but I'll see if I can figure something out.